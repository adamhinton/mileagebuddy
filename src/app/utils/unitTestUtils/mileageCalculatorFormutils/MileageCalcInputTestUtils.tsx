// README
// This is utils for testing the MileageCalcFormTextInput and MileageCalcFormNumInput components
// They're very similar, so share much utils logic
// This is for mocking relevant functions, and creating a wrapper component for testing

import MileageCalcFormNumInput from "@/app/calculator/CalculatorFormComponents/MileageCalcFormNumberInput";
import MileageCalcFormTextInput from "@/app/calculator/CalculatorFormComponents/MileageCalcFormTextInput";
import { Path } from "react-hook-form";
import { z } from "zod";
import { Vehicle_For_db_POST } from "../../server/types/VehicleTypes/POSTVehicleTypes";

const mockRegister = jest.fn().mockReturnValue({
	onChange: jest.fn(),
	onBlur: jest.fn(),
	name: "test",
});

export const MileageInputNumTestWrapper = ({
	path,
	error,
	schema,
}: {
	path: Path<Vehicle_For_db_POST>;
	error?: string;
	schema: z.ZodNumber;
}) => {
	return (
		<MileageCalcFormNumInput
			registerFn={mockRegister}
			path={path}
			error={error}
			subSchema={schema}
		/>
	);
};

export const MileageInputTextTestWrapper = ({
	path,
	error,
	schema,
}: {
	path: Path<Vehicle_For_db_POST>;
	error?: string;
	schema: z.ZodString;
}) => {
	return (
		<MileageCalcFormTextInput
			registerFn={mockRegister}
			path={path}
			error={error}
			subSchema={schema}
		/>
	);
};
