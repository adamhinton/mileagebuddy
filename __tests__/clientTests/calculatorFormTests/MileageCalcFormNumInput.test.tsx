// README
// This is test for the MileageCalcFormNumInput component (obviously)
// It shares a lot of logic with MileageCalcFormTextInput, so these two tests share a utils file and many of the tests will look similar

import { MileageInputNumTestWrapper } from "@/app/utils/unitTestUtils/mileageCalculatorFormutils/MileageCalcInputTestUtils";
import { render, screen } from "@testing-library/react";
// Importing this is essential for stuff like toBeVisible()
import "@testing-library/jest-dom";
import { z } from "zod";

describe("Sanity check", () => {
	it("should pass", () => {
		expect(1 + 1).toBe(2);
	});
});

// This test is run for both components
describe("Common Input Behavior", () => {
	it("Renders without errors", () => {
		render(
			<MileageInputNumTestWrapper
				path="yearlyMaintenanceCosts.batteries"
				schema={z.number().describe("My Vehicle Number")}
			/>
		);
		const label = screen.getByText("My Vehicle Number");
		const input = screen.getByTestId("yearlyMaintenanceCosts.batteries-input");

		expect(input).toBeVisible();
		expect(label).toBeVisible();
	});

	it("renders with label from path when schema has no description", () => {
		render(
			<MileageInputNumTestWrapper
				path="gasVehicleData.milesPerGallonCity"
				schema={z.number()}
			/>
		);
		const label = screen.getByText("milesPerGallonCity");

		expect(label).toBeVisible();
	});

	it("shows required asterisk for required fields", () => {
		render(
			<MileageInputNumTestWrapper
				path="gasVehicleData.milesPerGallonCity"
				// Defaults to not optional
				schema={z.number().nonnegative()}
			/>
		);
		const label = screen.getByText("*");

		expect(label).toBeVisible();
	});

	it("displays error message when error prop provided", () => {
		render(
			<MileageInputNumTestWrapper
				path="gasVehicleData.milesPerGallonCity"
				schema={z.number()}
				error="This is an error"
			/>
		);
		const error = screen.getByText("This is an error");

		expect(error).toBeVisible();
		expect(error).toHaveTextContent("This is an error");
	});

	it("hides error message when error prop is undefined", () => {
		render(
			<MileageInputNumTestWrapper
				path="gasVehicleData.milesPerGallonCity"
				schema={z.number()}
			/>
		);
		const error = screen.queryByTestId(
			"gasVehicleData.milesPerGallonCity-error"
		);

		expect(error).toBeNull();
	});

	it("Toggles error message when re-rendered with different error prop", () => {
		const { rerender } = render(
			<MileageInputNumTestWrapper
				path="gasVehicleData.milesPerGallonCity"
				schema={z.number()}
				error="This is an error"
			/>
		);
		const error = screen.getByText("This is an error");
		expect(error).toBeVisible();

		// Re-render without error
		rerender(
			<MileageInputNumTestWrapper
				path="gasVehicleData.milesPerGallonCity"
				schema={z.number()}
			/>
		);
		const noError = screen.queryByTestId(
			"gasVehicleData.milesPerGallonCity-error"
		);
		expect(noError).toBeNull();

		// Re-render with different error
		rerender(
			<MileageInputNumTestWrapper
				path="gasVehicleData.milesPerGallonCity"
				schema={z.number()}
				error="A different error"
			/>
		);
		const newError = screen.getByText("A different error");
		expect(newError).toBeVisible();
	});
});

describe("Number Input Specific", () => {
	it("enforces min value of 0 for nonnegative fields", () => {
		render(
			<MileageInputNumTestWrapper
				path="electricVehicleData.costPerCharge"
				schema={z.number().nonnegative()}
			/>
		);
		const input = screen.getByTestId("electricVehicleData.costPerCharge-input");

		expect(input).toHaveAttribute("min", "0");
	});

	it("enforces max value from schema", () => {
		render(
			<MileageInputNumTestWrapper
				path="yearlyMaintenanceCosts.batteries"
				schema={z.number().max(100)}
			/>
		);
		const input = screen.getByTestId("yearlyMaintenanceCosts.batteries-input");

		expect(input).toHaveAttribute("max", "100");
	});

	// Allows user to input two decimal places for gas/charging cost fields
	// All other fields increment by the dollar
	it('uses step="0.01" for gas/charging cost fields', () => {
		render(
			<MileageInputNumTestWrapper
				path="gasVehicleData.gasCostPerGallon"
				schema={z.number()}
			/>
		);
		const input = screen.getByTestId("gasVehicleData.gasCostPerGallon-input");

		expect(input).toHaveAttribute("step", "0.01");

		render(
			<MileageInputNumTestWrapper
				path="electricVehicleData.costPerCharge"
				schema={z.number()}
			/>
		);
		const input2 = screen.getByTestId(
			"electricVehicleData.costPerCharge-input"
		);

		expect(input2).toHaveAttribute("step", "0.01");
	});
	// it('uses step="1" for other number fields', () => {});
	// it("converts input value to number type", () => {});
});
