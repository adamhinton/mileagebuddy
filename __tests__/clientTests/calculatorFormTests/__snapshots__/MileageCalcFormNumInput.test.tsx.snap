// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Tests matches snapshot for basic number input from 4.30.2025 1`] = `
<div>
  <div
    class="mb-4"
  >
    <label
      class="mileage-calc-form-label"
      data-testid="yearlyMaintenanceCosts.batteries-label"
      for="yearlyMaintenanceCosts.batteries"
    >
      My Vehicle Number
      <span
        class="mileage-calc-form-required-asterisk"
      >
        *
      </span>
    </label>
    <div
      class="relative rounded-md shadow-sm"
    >
      <input
        class="mileage-calc-form-number-field "
        data-testid="yearlyMaintenanceCosts.batteries-input"
        id="yearlyMaintenanceCosts.batteries"
        min="0"
        name="test"
        required=""
        step="1"
        type="number"
      />
    </div>
  </div>
</div>
`;

exports[`Snapshot Tests matches snapshot for decimal step input from 4.30.2025 1`] = `
<div>
  <div
    class="mb-4"
  >
    <label
      class="mileage-calc-form-label"
      data-testid="gasVehicleData.gasCostPerGallon-label"
      for="gasVehicleData.gasCostPerGallon"
    >
      gasCostPerGallon
      <span
        class="mileage-calc-form-required-asterisk"
      >
        *
      </span>
    </label>
    <div
      class="relative rounded-md shadow-sm"
    >
      <input
        class="mileage-calc-form-number-field "
        data-testid="gasVehicleData.gasCostPerGallon-input"
        id="gasVehicleData.gasCostPerGallon"
        min="0"
        name="test"
        required=""
        step="0.01"
        type="number"
      />
    </div>
  </div>
</div>
`;

exports[`Snapshot Tests matches snapshot for number input with error from 4.30.2025 1`] = `
<div>
  <div
    class="mb-4"
  >
    <label
      class="mileage-calc-form-label"
      data-testid="gasVehicleData.milesPerGallonCity-label"
      for="gasVehicleData.milesPerGallonCity"
    >
      milesPerGallonCity
      <span
        class="mileage-calc-form-required-asterisk"
      >
        *
      </span>
    </label>
    <div
      class="relative rounded-md shadow-sm"
    >
      <input
        aria-describedby="gasVehicleData.milesPerGallonCity-error"
        class="mileage-calc-form-number-field mileage-calc-form-number-field-error"
        data-testid="gasVehicleData.milesPerGallonCity-input"
        id="gasVehicleData.milesPerGallonCity"
        min="0"
        name="test"
        required=""
        step="1"
        type="number"
      />
    </div>
    <p
      class="mileage-calc-form-error-message"
      data-testid="gasVehicleData.milesPerGallonCity-error"
      id="gasVehicleData.milesPerGallonCity-error"
    >
      <svg
        class="mileage-calc-form-error-icon"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M10 18a8 8 0 100-16 8 8 0 000 16zM8.28 7.22a.75.75 0 00-1.06 1.06L8.94 10l-1.72 1.72a.75.75 0 101.06 1.06L10 11.06l1.72 1.72a.75.75 0 101.06-1.06L11.06 10l1.72-1.72a.75.75 0 00-1.06-1.06L10 8.94 8.28 7.22z"
          fill-rule="evenodd"
        />
      </svg>
      This is an error message
    </p>
  </div>
</div>
`;

exports[`Snapshot Tests matches snapshot for required number field from 4.30.2025 1`] = `
<div>
  <div
    class="mb-4"
  >
    <label
      class="mileage-calc-form-label"
      data-testid="gasVehicleData.milesPerGallonCity-label"
      for="gasVehicleData.milesPerGallonCity"
    >
      milesPerGallonCity
      <span
        class="mileage-calc-form-required-asterisk"
      >
        *
      </span>
    </label>
    <div
      class="relative rounded-md shadow-sm"
    >
      <input
        class="mileage-calc-form-number-field "
        data-testid="gasVehicleData.milesPerGallonCity-input"
        id="gasVehicleData.milesPerGallonCity"
        min="0"
        name="test"
        required=""
        step="1"
        type="number"
      />
    </div>
  </div>
</div>
`;
